collect_app/src/main/java/org/odk/collect/android/formmanagement/FormUpdateChecker.kt

```diff
@@ -0,0 +1,100 @@
+import org.odk.collect.android.storage.StoragePathProvider
+import org.odk.collect.android.storage.StorageSubdirectory
+import org.odk.collect.android.utilities.FormsDirDiskFormsSynchronizer
+import org.odk.collect.android.utilities.FormsRepositoryProvider
+import org.odk.collect.android.utilities.TranslationHandler
+import org.odk.collect.forms.FormSourceException
+import java.io.File
+import java.util.stream.Collectors
+
+class FormUpdateChecker(
+    private val context: Context,
+    private val notifier: Notifier,
+    private val analytics: Analytics,
+    private val changeLock: ChangeLock,
+    private val storagePathProvider: StoragePathProvider,
+    private val settingsProvider: SettingsProvider,
+    private val formsRepositoryProvider: FormsRepositoryProvider,
+    private val formSourceProvider: FormSourceProvider
+) {
+
+    fun checkForUpdates(projectId: String): Boolean {
+        val formsDirPath = formsDir(projectId)
+        val formsRepository = formsRepository(projectId)
+        val formSource = formSource(projectId)
+        val diskFormsSynchronizer: DiskFormsSynchronizer = FormsDirDiskFormsSynchronizer(
+            formsRepository,
+            formsDirPath
+        )
+
+        val serverFormsDetailsFetcher = ServerFormsDetailsFetcher(
+            formsRepository,
+            formSource,
+            diskFormsSynchronizer
+        )
+
+        val cacheDirPath: String = formsCacheDir(projectId)
+        val formDownloader = ServerFormDownloader(
+            formSource,
+            formsRepository,
+            File(cacheDirPath),
+            formsDirPath,
+            FormMetadataParser(ReferenceManager.instance()),
+            analytics
+        )
+
+        val generalSettings = generalSettings(projectId)
+
+        return try {
+            val serverForms: List<ServerFormDetails> = serverFormsDetailsFetcher.fetchFormDetails()
+            val updatedForms =
+                serverForms.stream().filter { obj: ServerFormDetails -> obj.isUpdated }
+                    .collect(Collectors.toList())
+            if (updatedForms.isNotEmpty()) {
+                if (generalSettings.getBoolean(GeneralKeys.KEY_AUTOMATIC_UPDATE)) {
+                    val formUpdateDownloader = FormUpdateDownloader()
+                    val results = formUpdateDownloader.downloadUpdates(
+                        updatedForms,
+                        changeLock,
+                        formDownloader,
+                        TranslationHandler.getString(context, R.string.success),
+                        TranslationHandler.getString(context, R.string.failure)
+                    )
+
+                    notifier.onUpdatesDownloaded(results)
+                } else {
+                    notifier.onUpdatesAvailable(updatedForms)
+                }
+            }
+
+            context.contentResolver.notifyChange(FormsProviderAPI.CONTENT_URI, null)
+            true
+        } catch (e: FormSourceException) {
+            true
```

Comment:
Shouldn't it return false if something goes wrong?

Reply:
Yeah, this confused me a little. The existing code returns `true` here though. I think it's because returning `false` in the `TaskSpec` would make the job run again which we don't do for auto update - even if there is a failure we count the job itself as "successful". It is confusing to have it here though. I'll rework this so that `checkForUpdates` is void as it always returns `true` at the moment.